prepare for release

WaffleLapkin opened the pull request:
- change impl details & make Entry skip on drop
- add rustfmt.toml
- pre-release things

WaffleLapkin committed:
change impl details & make Entry skip on drop

For some reason, in my benches this impl performs worse. Then the previous.
But it feels like it will be easier to add backwords iteration to this impl,
so I'll probably stick with it.

Without skip on drop, `Entry` is *very* easy to miss use.

WaffleLapkin committed:
add rustfmt.toml

WaffleLapkin committed:
pre-release things

WaffleLapkin committed:
Merge branch 'master' into prepare_for_release

WaffleLapkin committed:
fix netlify doc build

WaffleLapkin committed:
increace minimum rustc version in CI

WaffleLapkin committed:
fix netlify docs build (again)

:facepalm:

WaffleLapkin committed:
fix CI

AnthonyMikh commented on src/lib.rs:
```diff
@@ -124,15 +124,75 @@ use core::ptr;
+    // Derived from `vec`
+    /// Points to the start of the buffer
+    #[cfg(all(x, not(x)))]
+    _start: NonNull<T>,
```
What's the point of having this field?

AnthonyMikh commented on src/lib.rs:
```diff
@@ -124,15 +124,75 @@ use core::ptr;
+    _start: NonNull<T>,
+
+    /// Points to the first empty 'slot'. If no items were removed, then points
+    /// to the same item as `curr`. Otherwise pints to the first empty
```
```suggestion
    /// to the same item as `curr`. Otherwise points to the first empty
```

WaffleLapkin committed:
revert impl

Turns out initial impl was both simpler&safer than anything I could comeup afterwords :/

AnthonyMikh commented on src/lib.rs:
```diff
@@ -125,14 +125,60 @@ use core::ptr;
     // - vec.capacity() >= len
-    // - vec[..slot] is initialized (note: dye to safety guarantees vec.len is
-    //   set to 0, so this is not 'real' indexing)
+    // - vec[..slot] is initialized (note: dye to safety guarantees vec.len is set to 0, so this is
```
> dye to

due to

AnthonyMikh commented on src/lib.rs:
```diff
@@ -209,71 +276,77 @@ impl<'a, T> Removing<'a, T> {
+        unsafe {
+            // ## Safety
+            //
+            //`Entry` type invatiants ensure that `curr` ptr is valid.
```
```suggestion
            //`Entry` type invariants ensure that `curr` ptr is valid.
```

AnthonyMikh commented on src/lib.rs:
```diff
@@ -209,71 +276,77 @@ impl<'a, T> Removing<'a, T> {
+        unsafe {
+            // ## Safety
+            //
+            //`Entry` type invatiants ensure that `curr` ptr is valid.
```
```suggestion
            //`Entry` type invariants ensure that `curr` ptr is valid.
```

AnthonyMikh commented on src/lib.rs:
```diff
@@ -328,12 +422,40 @@ impl<T> VecExt<T> for Vec<T> {
     /// Returns non-copy type that can help miri detect safety bugs
-    fn f(i: i32) -> impl Debug + Eq {
-        i.to_string()
+    fn f(i: i32) -> impl Copy + Debug + Eq + PartialOrd<i32> + Rem<i32, Output = i32> {
```
> Returns non-copy type

This comment seems outdated.

WaffleLapkin committed:
remove Copy impl from tests

WaffleLapkin committed:
fix typo

WaffleLapkin committed:
Apply suggestions from code review

Co-authored-by: AnthonyMikh <anthony.mikh@yandex.ru>

WaffleLapkin committed:
downgrade msrv version back to 1.36.0

WaffleLapkin committed:
document & better test forget behaviour

AnthonyMikh commented on src/lib.rs:
```diff
@@ -57,12 +57,15 @@
-/// Essentially overhead of the process is one `ptr::copy` for each [`skip`]ed
-/// element + one `ptr::copy` for the not yielded tail.
+/// Essentially overhead of the process is one `ptr::copy` for each ignored
+/// element element + one `ptr::copy` for the not yielded tail.
```
"element" is doubled

AnthonyMikh commented on src/lib.rs:
```diff
@@ -125,14 +125,60 @@ use core::ptr;
     slot: usize,
     curr: usize,
     len: usize,
+    /* Example of how this works, step-by-step:
```
I think it makes sense to put this explanatory comment outside of struct definition

AnthonyMikh commented on src/lib.rs:
```diff
@@ -308,6 +414,22 @@ impl<T> Entry<'_, '_, T> {
 
+impl<T> Drop for Entry<'_, '_, T> {
+    fn drop(&mut self) {
+        // Skips element swaping it with the first empty slot.
```
```suggestion
        // Skips element swapping it with the first empty slot.
```

WaffleLapkin committed:
clean documentation of Removing's forget behavior

WaffleLapkin committed:
fix typo & fotmatting

WaffleLapkin committed:
Update src/lib.rs

Co-authored-by: AnthonyMikh <anthony.mikh@yandex.ru>
